eeps/eep-0066.md

```diff
@@ -0,0 +1,438 @@
+as the second or later characters in a variable or an atom.
+In short ISO [Latin-1][] letters, digits, «`_`» and «`@`».
+The [Sigil Type][] may be empty.
+
+The Sigil Type defines how the [Sigil][] syntactic sugar
+shall be interpreted.  The suggested Sigil Types are:
+
+* «»: the Vanilla [Sigil][].
+
+  Creates an Erlang `unicode:unicode_binary()`.
+  It is a string represented as a UTF-8 encoded binary,
+  equivalent to applying `unicode:characters_to_binary/1`
+  on the [String Content][].  The [String Delimiters][]
+  and escape characters work as for regular strings,
+  triple-quoted strings, or quoted atoms in Erlang.
+
+  So «`~"abc\d"`» is equivalent to «`<<"abc\d"/utf8>>`», and
+  «`~'abc"d'`» is equivalent to «`<<"abc\"d"/utf8>>`».
+
+  «`~"`» would work as «`~s"`» and «`~"""`» would work
+  as «`~S"""`» below, regarding escape characters.
+
+  A simple way to create strings as UTF-8 binaries is supposedly
+  the first and most desired missing string feature in Erlang.
+  This sigil does just that and has no other features.
+
+* «`s`»: [string in Elixir][4].
+
+  Creates an Erlang `unicode:unicode_binary()`, handling
+  escape characters in the string content.  How other features
+  like string interpolation would work is still an open question.
+  
+  Escape characters and other features are the same regardless
+  of which [String Delimiters][] that are used.
+
+* «`S`»: [string in Elixir][4], verbatim.
+
+  Creates an Erlang `unicode:unicode_binary()`, with verbatim
+  string content in that only the [end delimiter][] character
+  can be escaped with a «`\`» character.  How other features
+  like string interpolation would work is still an open question.
+
+  Escape characters and other features are the same regardless
+  of which [String Delimiters][] that are used, except that
+  between triple-quote delimiters according to [EEP 64][]
+  there is no end delimiter character to escape.
+
+* «`r`»: regular expression.
+
+  Creates a term `{RE::unicode:charlist(),Flags::[unicode:latin1_char()]}`
+  that is an uncompiled regular expression with compile flags,
+  suitable for functions in the `re` module.  The `RE` element is
+  the [String Content][], and the `Flags` element is the [Sigil Suffix][].
+
+  See the [Regular Expressions][] section about the reasoning
+  behind this proposed term type.
+
+  Within the [String Content][], character escape sequences are handled
+  according to the regular expression rules, and in addition to them
+  the [end delimiter][] character can be escaped with a «`\`» character.
+  Between triple-quote delimiters according to [EEP 64][]
+  there is no end delimiter character to escape.
+
+  The main advantage of a regular expression [Sigil][] is to avoid
+  the additional escaping of «'\'» that regular erlang strings add.
+
+  Today: `re:run(Subject, "^[ \\t]*\\[a-z]*\\\\s+", [caseless,unicode])`
+
+  Sigil: `re:run(Subject, ~r"^[ \t]*\[a-z]*\\s+"iu)`
```

Comment:
The more I think about it the less I like the idea of having a specific sigil for regular expressions, what is the main benefit over `re:run(Subject, ~S"^[ \t]*\[a-z]*\\s+", [i, u])`?

Reply:
Yes, I think it would be valuable if we could compile them at load-time. At time point the value is relatively limited.
Unless the representation of `~r` is left unspecified, such that it would be backwards-compatible to later compile them at load-time in a subsequent release.

Reply:
I'd say possible `re` integration, future pre-compilation, and the possibility to handle the regular expression and its compilation flags as one unit (+ less punctuation characters for the compile flags).

Reply:
If we would take an intermediate step to not add `~r` now, then people will start using `~S` for regular expressions, and then want to switch to using `~r` when that is introduced.

I don't know if it will be annoying to have stepped intermediate step...

Reply:
I think `~r` would make more sense if it was tied to an operator like the Perl match operator. Maybe use multiple sigils, like `Subject <> ~m"..."` to match (or `~m/.../`), `Subject <> ~s".../..."` to replace. The pre compilation alone that `~r` could provide is not super interesting.

Reply:
I do not think we want to tie regular expressions into the base language that hard, as an operator would do. Erlang should not try to be Perl.

If we would have a match operator, then there would be no possibility to e.g strip the `re` module from Stdlib.  It would irreversibly be a part of the Erlang language. I think it should be a library.  But some language support like this would be more OK.

Reply:
The pre compilation discussed here would also tie the Erlang language to a regular expression implementation though. I do not think an operator is too different. Just have to devise a way to make the implementation pluggable instead of relying directly on `re`.

The operator doesn't have to specify the way the regular expression implementation behaves, just what happens when there is/isn't a match. Then users can refer to the implementation they use for the details of the syntax and behavior.

I'm not saying this should be just like Perl, as Perl goes far beyond this and not in a good way IMO. But I don't see much point in having `~r` if it's going to either be like `~S` or just provides some pre compilation magic on top.

Reply:
> Maybe use multiple sigils, like Subject <> ~m"..." to match (or ~m/.../), Subject <> ~s".../..." to replace. 

I am not the biggest fan of making the result of the operation based on the sigil. Maybe `Subject <> ~m"foo"` means match, maybe `Subject <> ~s"foo/bar"` means replace, but you can also write `Subject <> AnotherVar`, and now I have no idea what it is going to return without knowing the exact type of `AnotherVar`.

I would also say that, from experience, match itself is not terribly useful. In Perl, `=~` sets magic variables with the regex captures, so it works both as `re:run/2` or also a quick check to see if something matched or not.

I believe no one is advocating for magic variables in Erlang, so the only option is to use `=~` (or similar) to check if something matched or not. We do have this operator in Elixir and, to be honest, the operator is almost exclusively used in tests, and that's it. Because in actual code, you most often don't want to just check if a regex matches, but also extract something from it. If that's the primary benefit, does it justify the changes to the language?

Reply:
In actual code I almost never see or use regular expressions, so I think the question applies to any language feature related to them. Perhaps the benefits of `~r` or equivalent are just too small to matter.

About the operator obviously that wasn't a fully thought out idea, and sigils might not be the right way to indicate match/replace, definitely. But I didn't think of `Subject <> ~m"foo"` as match-only but as something that would return the captures if any were defined.

Anyway details aside I think that my point that `~r` on its own not being super useful still stands. If the language allows defining regular expressions it should also allow using them directly otherwise they're no better than regular strings.

Reply:
That's a chicken-egg issue though. Regexes are very awkward to write today, so it is understandable there are rarely used. I would say we should at least make them easier to write.

Reply:
I'm not disagreeing with `~S` which would make them easy to write already. The verbatim stuff is great. I'm just saying if you introduce something specific to regular expressions it has to either come with the relevant features or that those features are intended to be introduced at a later time. Stopping at `~S` + maybe a pre compilation later is not interesting IMO.

Reply:
@essen wrote:
> The pre compilation discussed here would also tie the Erlang language to a regular expression implementation though. I do not think an operator is too different. Just have to devise a way to make the implementation pluggable instead of relying directly on re.

An operator would require that there *is* an regular expression engine in the system or else the syntax does not make sense.

The current suggestion is that `~r"abc\d"ef` translates to `{re,"abc\\d","ef"}`, which is a pluggable interface, and fairly general.  In most other interfaces I know of there is the expression and the compile flags, both as strings.  Then a regular expression module should accept such a term for the combined regular expression and its compile flags.

If there is such a module, the code loader can be made to pre-compile regular expression literals used to call functions in the regular expression module.

> But I don't see much point in having ~r if it's going to either be like ~S or just provides some pre compilation magic on top.

The other benefit is to be able to combine the regular expression and its compile flags into one unit, in a way familiar to other regular expression interfaces such as sed, awk, perl, ...